### PR TITLE
Build assets using Webpack

### PR DESCRIPTION
**Pull request by SeinopSys**
### Before you begin

* I understand my contributions may be rejected for any reason
* I understand my contributions are for the benefit of Derpibooru and/or the Philomena software
* I understand my contributions are licensed under the GNU AGPLv3

- [x] I understand all of the above

---

Replaced the build system for frontend assets with Webpack.

### CODE REVIEW

File: assets/css/common/_base.scss [hunk 1]
```diff
@@ -11,11 +11,12 @@
 // Because FA is a SPECIAL SNOWFLAKE.
 $fa-font-path: "/fonts";
 
-@import "../../node_modules/@fortawesome/fontawesome-free/scss/fontawesome.scss";
-@import "../../node_modules/@fortawesome/fontawesome-free/scss/solid.scss";
-@import "../../node_modules/@fortawesome/fontawesome-free/scss/regular.scss";
-@import "../../node_modules/@fortawesome/fontawesome-free/scss/brands.scss";
-@import "../../node_modules/normalize-scss/sass/normalize/import-now";
+$fa-font-path: '~@fortawesome/fontawesome-free/webfonts';
```
Comment (liamwhite) [inline assets/css/common/_base.scss:14]: Reassigning the same variable

File: assets/css/common/_base.scss [hunk 2]
```diff
@@ -380,23 +381,23 @@ span.stat {
...
-@import "views/galleries";
-@import "views/images";
-@import "views/pages";
-@import "views/polls";
-@import "views/posts";
-@import "views/profiles";
-@import "views/search";
-@import "views/stats";
-@import "views/tags";
+@import "../views/adverts";
```
Comment (liamwhite) [inline assets/css/common/_base.scss:384]: Is there no way to alter the import search path?

File: assets/js/app.js [hunk 3]
```diff
@@ -15,3 +15,7 @@ import './vendor/es6.polyfill';
 // Our code
 import './ujs';
 import './when-ready';
+
+import '../css/themes/default.scss';
```
Comment (liamwhite) [inline assets/js/app.js:19]: Maybe I'm just being pedantic, but this is _so_ dumb. Is there a way to do this without literally importing the CSS files into the scripts?

File: assets/package.json [hunk 4]
```diff
@@ -1,20 +1,33 @@
 {
   "scripts": {
-    "deploy": "brunch build --production",
-    "watch": "brunch watch --stdin"
+    "build": "cross-env NODE_ENV=production webpack",
```
Comment (liamwhite) [inline assets/package.json:3]: The name chosen is deploy [such as to be consistent with the Phoenix deployment guidelines](https://hexdocs.pm/phoenix/deployment.html#compiling-your-application-assets)

File: assets/webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
+const path = require('path');
```
Comment (liamwhite) [inline assets/webpack.config.js:1]: [For reference, here's how Phoenix does it upstream](https://github.com/phoenixframework/phoenix/blob/v1.5.3/installer/templates/phx_assets/webpack.config.js)

**Comment by liamwhite** [comment]
I had some luck decreasing the output bundle size down to 68K by
1. using webpack-rollup-loader
2. setting `runtimeChunk: true` to disable the webpack loader chunk at the top
3. removing the reference to setImmediate in promise.polyfill.js to get rid of a massive chunk of node-specific code that gets injected without any way to disable it (seriously???)

...but babel is still inserting numerous copies of rather large behavior polyfills and I don't know how to stop it from doing that.

**Comment by SeinopSys** [comment]
I thin the filesize increase might be because those polyfills were never inserted previously, `@babel/preset-env` wasn't even a dependency before. The supported environments [can be adjusted via `.babelrc`](https://babeljs.io/docs/en/babel-preset-env), I can give that a shot to make it match the browserlist passed to autoprefixer

File: assets/js/app.js (hunk 3 above)
Comment (SeinopSys) [reply assets/js/app.js:19]: I tried adding them as separate entrypoints in the config caused files to be written to the output that only had webpack's runtime in it. You somehow have to tell webpack about these files' existence and this is the only way I know to do that.

**Comment by SeinopSys** [comment]
@liamwhite Please have another look, I made the adjustments you suggested. These are the new prod build results on myend.

| size | file |
| ---- | ---- |
| js/app.js | 59.9 KiB |
| css/dark.css | 102 KiB |
| css/default.css | 101 KiB |
| css/red.css | 102 KiB |

**Comment by liamwhite** [comment]
This is better in terms of filesize (and gets even better if you remove the setImmediate fill in promise.polyfill.js), but I'm seeing ES arrow functions and for..of loops in the output, which do not belong there.

**Comment by SeinopSys** [comment]
As a quick sanity check I also just tried IE and apparently the fetch polyfill cannot find the promise polyfill. I'll dig into both of these issues.

**Comment by SeinopSys** [comment]
Removing the `setImmediate` broke the polyfill because it was using it, so I just changed the check to the property on the window reference that's passed as a parameter and that seemed to have the same size reduction effect.

File: assets/js/app.js (hunk 3 above)
Comment (SeinopSys) [reply assets/js/app.js:19]: I updated this to use `resolve.alias` but then `~` is required in front of the alias names, that's the best that can be done by the looks of it.

**Comment by liamwhite** [comment]
I just did this
```js
  // Use polyfill for setImmediate for performance gains
  function asap(fn) {
    setTimeoutFunc(fn, 1);
  };
```

**Comment by SeinopSys** [comment]
Updated.

**Pull request merged**